### PR TITLE
Support feature vector for cls

### PR DESCRIPTION
**Pull request by negvet**
### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/core/model/entity/classification.py [hunk 1]
```diff
@@ -622,7 +654,7 @@ def _export_parameters(self) -> dict[str, Any]:
         export_params["via_onnx"] = False
         export_params["input_size"] = self.image_size
         export_params["onnx_export_configuration"] = None
-        export_params["output_names"] = ["logits", "saliency_map"] if self.explain_mode else None
+        export_params["output_names"] = ["logits", "feature_vector", "saliency_map"] if self.explain_mode else None
```
Comment (GalyaZalesskaya) [inline src/otx/core/model/entity/classification.py]: What do you think about moving explain related parameters to src/otx/core/model/entity/classification.py?

ExplainableOTXClsModel ->
```
    @property
    def _export_parameters(self) -> dict[str, Any]:
        """Defines parameters required to export a particular model implementation."""
        export_params = super()._export_parameters
        export_params["output_names"] = ["logits", "feature_vector", "saliency_map"] if self.explain_mode else None
        return export_params
```

Comment (negvet) [reply src/otx/core/model/entity/classification.py]: I think it is possible, my logic was to put it near `export_params["onnx_export_configuration"]`. But I can move it up. Probably some other parameters also can be moved to the upper level class, @sovrasov?

Comment (sovrasov) [reply src/otx/core/model/entity/classification.py]: We have to add the parameters which make sense at the current level of hierarchy. In `ExplainableOTXClsModel` we already know what the outputs should be in explain mode, so we can move `export_params["output_names"]` there.

Other parameters here are implementation-specific (in out case that's mmpretrain), and I'd opt to leave them as is. Harim is working on torchvision wrapper, so we'd better keep parameters which could be not generalizable at lower levels.

**Review by sovrasov — DISMISSED** [review]

**Review by harimkang — DISMISSED** [review]

File: src/otx/algo/hooks/recording_forward_hook.py [hunk 2]
```diff
@@ -16,6 +16,24 @@
...
 
 
+def feature_vector_fn(feature_map: torch.Tensor | Sequence[torch.Tensor]) -> torch.Tensor:
+    """Generate the feature vector by average pooling feature maps."""
+    if isinstance(feature_map, (list, tuple)):
+        # aggregate feature maps from Feature Pyramid Network
+        feature_vectors = [torch.nn.functional.adaptive_avg_pool2d(f, (1, 1)) for f in feature_map]
+        for i in range(len(feature_vectors)):
+            batch, num_channels, *_ = feature_vectors[i].shape
+            feature_vectors[i] = feature_vectors[i].view(batch, num_channels)
```
Comment (vinnamkim) [inline src/otx/algo/hooks/recording_forward_hook.py]:
```suggestion
        feature_vectors = [
             # Spatially pooling and flatten, B x C x H x W => B x C'
             torch.nn.functional.adaptive_avg_pool2d(f, (1, 1)).flatten(start_dim=1)
             for f in feature_map
        ]
```

File: src/otx/algo/hooks/recording_forward_hook.py [hunk 3]
```diff
@@ -16,6 +16,24 @@
...
+
+        if len(feature_vectors) > 1:
+            return torch.cat(feature_vectors, 1)
+        return feature_vectors[0]
+
+    feature_vector = torch.nn.functional.adaptive_avg_pool2d(feature_map, (1, 1))
+    batch, num_channels, *_ = feature_vector.shape
+    return feature_vector.view(batch, num_channels)
```
Comment (vinnamkim) [inline src/otx/algo/hooks/recording_forward_hook.py]: Here as well like https://github.com/openvinotoolkit/training_extensions/pull/3047/files#r1515331667

File: src/otx/core/model/entity/classification.py [hunk 4]
```diff
@@ -673,7 +709,7 @@ def _customize_outputs(
             # Squeeze dim 4D => 3D, (1, num_classes, H, W) => (num_classes, H, W)
             predicted_s_maps = [out.saliency_map[0] for out in outputs]
 
-            predicted_f_vectors = [out.feature_vector for out in outputs]
+            predicted_f_vectors = [out.feature_vector[0] for out in outputs]
```
Comment (vinnamkim) [inline src/otx/core/model/entity/classification.py:713]: Would you add a comment here about the reason why we extract index = 0 from `out.feature_vector`?

Comment (negvet) [reply src/otx/core/model/entity/classification.py:713]: Sure, added

**Review by vinnamkim — APPROVED** [review]

**Review by sovrasov — APPROVED** [review]

**Review by GalyaZalesskaya — APPROVED** [review]

**Pull request merged**